### ISSUE TITLE
README instructions for migrating up and down

### ISSUE THREAD

**Issue by EliuX** [cocreators-ee/migrate-anything#3]
Hi guys, KUDOS for the library. I was trying to use it but there is no option available to migrate down or even list the current state of the migrations. Am I missing something?
Thanks in advance.

**Comment by lietu** [comment]
The `up` or `down` is automatic based on the filesystem state - e.g. if you switch between branches.

Let's say your current branch has the migrations

 - 001-setup.py
 - 002-user.py

Then switch to a branch with:

 - 001-setup.py
 - 002-user.py
 - 003-product.py

Then the new `03-product.py` migration would be run when you run the tool. The tool then stores the code in that file to the DB, so it can undo that when it is no longer in the filesystem.

When you delete that file manually, or switch back to the other branch without this new file and then run the tool, it will undo it. The sequencing is based on sorting the filenames, thus sequence numbers are useful.

There is no `up` or `down`, there is just `migrate to current state`, and the [examples](https://github.com/Lieturd/migrate-anything/tree/master/examples) have a few scripts to show how that works exactly, e.g. [examples/simple/test.sh](https://github.com/Lieturd/migrate-anything/blob/master/examples/simple/test.sh).

Right now you can just check the DB for current state, but it's something that could definitely be added to the tool at some point. If you add a feature request -issue, it's easier to remember that :)

**Comment by EliuX** [comment]
Greetings. One question. How can you execute the "down" logic of the files
you have erased?

**Comment by lietu** [comment]
I literally just told you, and pointed you to the examples.

```bash
migrate-anything migrations
```

There's just the one command. It runs the "down" logic of the files that are no longer present in the filesystem.

**Comment by EliuX** [comment]
I saw the examples you mentioned and I saw you said
>When you delete that file manually, or switch back to the other branch without this new file and then run the tool, it will undo it.

Then I dug into the README and I saw

>Any files that are missing from the fs but are in the DB will have their code loaded from the DB and their down() is executed - in reverse order

But what if I wanted to go `up` again? I already have those files deleted! 
In order to support this, instead of saving the migration, status could be just a track of the migration files and the time when they were executed. This way you could do a list and see where is the cursor of the migrations without having to delete them or persist them all. Have you considered this alternative?

**Comment by lietu** [comment]
That is what version control is for. You can e.g. `git stash` your files. This tools is specifically assuming you know what version control is, and how to use it.

This is a better solution than people manually managing which ones you want to `up`/`down` which requires manual effort and manual effort always leads to human error, but admittedly there is some value for testing the migrations to being able to manually `down` a specific one at will - so if you wish, you can add a feature request as well.

Basically what would satisfy your requirements should be something like:

```bash
migrate-anything migrations                  # Migrate to current state in FS
migrate-anything migrations list             # Show currently applied migrations
migrate-anything migrations down <filename>  # Undo the specific migration via code in DB
```

Feel free to submit a PR that adds those :)

However e.g. adding an `up` would promote bad practices that I don't think would be good without a very strong reason.

**Comment by EliuX** [comment]
Regarding 
>his is a better solution than people manually managing which ones you want to up/down which requires manual effort and manual effort always leads to human error

Well, erasing files or stashing them with a tool outside of this library would not require more effort than listing the migrations and going down to a particular one. In addition, running the `down` logic from an already persisted version is not intuitive, at least for me. Especially considering that I may want to tune the `down` logic if it fails. Also, consider that a `list` option would serve as a report of the status of the migrations. Right now you have to rely in a sorted representation of the files, using the local shell maybe. Also, mutating the state of files for getting a particular state is not a good practice for me: What if you want to run and down the migrations in a read-only file-system environment?

> so if you wish, you can add a feature request as well.
>However e.g. adding an up would promote bad practices that I don't think would be good without a very strong reason.

I really appreciate your feedback and in the first place that you have done the effort to create a migration library that is truly database-agnostic. Who would have thought that in Python there were no options for that? You saw the problem and created a solution, KUDOS! 
But if you consider that the current one is a better design (I respectfully disagree), I would rather be using it the way it is and respect your vision, because in the end you are the one(s)  putting effort into this.

**Comment by lietu** [comment]
> In addition, running the down logic from an already persisted version is not intuitive, at least for me.

Well the point is to not have 2 different ways to run `down` migrations, as that would again needlessly add complication, cognitive load, and chances for human error.

You do want to tune the `down` logic often I'm sure, and I don't see how it would be such a bad thing to require you to run the previously applied version to try it out before you can make changes and try again - I just can't imagine up a use-case that would require this.

Previously the way I've done this in practice was simply renaming the file e.g. `02-foo.py` to `02-foo2.py` and back, then when I'm done with it I'll just rename it to what it should be and then commit, but I can see this could be done in a more .. elegant manner 😄

If the `down` command was added, I don't think I would be particularly offended if it had an extra `--from-file` flag or similar that would ignore the source in the DB, if you think this is particularly useful. However, it does come with the downside of adding complications to a scenario which doesn't necessary need it.


> Also, consider that a list option would serve as a report of the status of the migrations.

Not sure what kind of "status" you're looking for here? Some kind of diffing the local files vs. what's in DB and showing `up-to-date`, `not applied`, `missing`, `changed` or so? This could well be useful, but I've found it largely unnecessary and thus it's not been implemented.

During development you typically should know pretty much what state you're in and should have tools to reset your state when necessary.

When deploying you shouldn't need to care what state it's in, just that it gets to the correct state - the one in the branch you deployed. This is basically the whole ideology shared by all deployment tools in the past 5 years or so - Docker, Salt Stack, Ansible, etc. - they are trying to ensure that your desired state goes in, without having to worry too much about the previous state (though in some cases you do need to take some special precautions).


> What if you want to run and down the migrations in a read-only file-system environment?

Seems like a bit contrived example that doesn't actually happen in the real world and I don't particularly feel the need to entertain such "what ifs".

You develop these files in your development environment, where your desktop either has a read-write file system or it's broken and you can't develop on it.

Afterwards you commit them to version control, push them off to your build pipelines etc. that do what needs to be done to ensure the right state ends up on the servers via e.g. Docker build (in which case you don't need write access to delete the files, the new container will just not have the obsolete files), and then execute `migrate-anything migrate`. Or you just run `git pull -u && migrate-anything migrate` or whatever to update the local state, requiring write access to the fs anyway.

As far as I can see, the best solution to this problem is to plan your setup so that you don't need to do it that way.


> if you consider that the current one is a better design

Of course I'm willing to listen to feedback, I'm far from perfect and my opinions are regularly proven wrong 🙂

I however do find it easier to consider these options when formulated in clear use-cases where the current solution is lacking without hypothetical scenarios, then I can more easily say if the tool is simply not what you're looking for and you might want to look for / build another tool, or maybe `migrate-anything` can be improved, or maybe there's some middle-ground that we can come up with that fulfills both your needs, and doesn't stray too far away from the original vision.

**Comment by lietu** [comment]
Also, in case I didn't make it clear before - I really do appreciate the feedback! I'm also happy to see there are others who can appreciate a tool like this.

I'll definitely be thinking on this a bit and might add a few issues myself on the subject to clarify the thoughts in clear actionable items.

**Comment by incorvia** [comment]
This definitely really confused me.. had to look at the examples and read it a few times before I understood how this worked. I work on Rails primarily so perhaps I just expected this to work the Rails way. The idea of deleting the file just to run the down migration does seem burdensome but maybe it's better and I just don't see it yet. If I make a mistake in my up migration and I want to roll it back, edit it, then re-run the migration.. I need to:

1. Delete the file or checkout a branch where the file doesn't exist, or stash the changes..
2. Re-run the migrations
3. Then either reinsert the file or checkout the branch where it does exist, or pop the stash..
4. Edit up migration
5. Re-run

I think for those of us that are used to simply:

1. Run down command
2. Edit
3. Run up command

On first inspection it seems like a lot of extra work.. I think, also, while I see your point about wanting it to 'just work' I think there's also a philosophy that could be taken which is that if I'm going to migrate the database I should need to explicitly run a command.. 

Anyways just adding my two cents.. will give this a try as is and if it doesn't work for me in the long run may just fork.

**Comment by incorvia** [comment]
Actually just ran into an issue where I had an error in my down migration but not in my up migration.. I stashed my changes to rollback, ran my migration and it errored, now I need to figure out how to fix this..

**Comment by lietu** [comment]
What I do is edit and rename files. The point with "just working" is for the people after you, they have to just switch the branch and run the tool, the tool will figure out the state you should be in, and run down and up as necessary.

**Comment by incorvia** [comment]
Thanks for responding @lietu. I'm trying to understand..  Just to clarify, if you've run the up but the down is 'bad' and the down is in the database, the recommended way to 'fix the bad down that is in the database' is to 'edit and rename the file'? Is this 'instead of deleting it' ?

**Comment by lietu** [comment]
No. Instead of

> 1. Delete the file or checkout a branch where the file doesn't exist, or stash the changes..
> 2. Re-run the migrations
> 3. Then either reinsert the file or checkout the branch where it does exist, or pop the stash..
> 4. Edit up migration
> 5. Re-run

1. Edit and rename the file
2. Re-run

I expected that by that time you had figured that if you've stored bad state in your database, you can delete it from the database.

**Comment by lietu** [comment]
Anyway, I don't see anything particularly useful discussion here, this is a zombie thread trying to be resurrected to unrelated things and people are likely getting pinged for it. If you have improvement ideas or PRs you can submit new ones.

**Comment by lietu** [comment]
Resolved by #4
